classes/various/extInteger.sc

```diff
@@ -0,0 +1,45 @@
+/*
+This file is part of MathLib, a Quark for SuperCollider:
+
+<https://github.com/supercollider-quarks/MathLib>
+
+You should have received a copy of the GNU General Public License v3 along with
+this distribution. If not, see:
+
+<https://www.gnu.org/licenses/gpl-3.0.txt>.
+*/
+
+
+//---------------------------------------------------------------------
+//	MathLib: Some mathematical extensions to SuperCollider.
+//
+// 	Extension: Integer
+//
+//---------------------------------------------------------------------
+
+
+/* J Anderson, 2019 */
++ Integer {
+
+	squareOf {
```

Comment:
I find the name a bit misleading. I read: `9.squareOf`, or `squareOf(9)`, I'd expect the result to be, well, the square of 9. Maybe `sqrtInt`? Btw. can the algorithm be generalized to the nth root? Then it would be `root`.

Reply:
Looks like the nth root is a bit more involved. Here's what it looks like for floats:

```supercollider
(
// Find *root* of *num* within *eps* precision
// https://en.wikipedia.org/wiki/Nth_root_algorithm
// eps of 0.001 seems quite good, diminishing returns beyond that
~rootOf = { | num, root, eps = 0.001|
	var cnt = 0, exp = 0, delta = inf;
	var two_exp = 1;
	var guess, rt_recip, rt_m1;

	// debug: if it loops this many times, something went wrong!
	var timeout = 1000; // remove after testing

	while({ two_exp < num }, {
		exp = exp + 1;
		two_exp = two_exp * 2;
	});

	// Initialize iteration with best-guess.
	// There are different approaches to scaling this:
	// exp/3 :https://www.boost.org/doc/libs/1_62_0/libs/math/doc/html/math_toolkit/roots/root_finding_examples/multiprecision_root.html
	// exp/n :https://www.boost.org/doc/libs/1_62_0/libs/math/doc/html/math_toolkit/roots/root_finding_examples/nth_root.html
	// exp*2/n found empirically : https://gist.github.com/mtmccrea/e6eab10ee4f6041e15b2a047aa799dfe
	guess = exp * 2 / root;

	rt_recip = root.reciprocal;
	rt_m1 = root - 1;
	while({ delta.abs > eps and: {cnt < timeout} }, {
		delta = rt_recip * (( num / guess.pow(rt_m1)) - guess);
		guess = guess + delta;
		cnt = cnt + 1;
	});

	if(cnt == timeout) {
		warn(format("Timed out: exceeded % iterations.", timeout))
	};

	guess // return
}
)

(
var num = 8320;
var root = 8;
var res = ~rootOf.(num, root, eps: 0.001);
postf("approx:\t%\ndiff:\t%\n", 
	res.pow(root), num - res.pow(root));
nil
)
```
The algorithm is found here:
https://en.wikipedia.org/wiki/Nth_root_algorithm

With some more involved testing and introspection here:
https://gist.github.com/mtmccrea/e6eab10ee4f6041e15b2a047aa799dfe

Reply:
good! Because it is more complex, it should be added as a separate method. E.g. it could be added as integerRootN or something similar.

Reply:
@mtmccrea this algorithm is interesting...

...but it isn't actually solving the problem I'm trying to nail. I think it is probably best viewed as a "quickie" approximation for something we already have access to.

If we need the n-th root of num, this will quickly do the job using builtins:

`~myNum.pow(~myRoot.reciprocal)`

Note: I haven't done any checking of speed to see how this compares to your approximation.

-----
So, my thoughts are:

- At the moment users do have access to the returned numerics, there may just be a question as to efficiency.
- My specific application (finding HOA order), requires knowing if an `Integer` (square) root can or can't be found.
- If a `Float` method were to be added, we'd probably like to have similar behavior and name given an `Integer` method.
- The algorithm in my [below comment](https://github.com/supercollider-quarks/MathLib/pull/23#issuecomment-716876026) solves my immediate problem, _and_ expands to finding n-th roots.